### ISSUE TITLE
Invalid response when uploading picture

### ISSUE THREAD

**Issue by sbertix** [TheM4hd1/SwiftyInsta#106]
_Originally posted by @trentona in https://github.com/TheM4hd1/SwiftyInsta/issues/105#issuecomment-535255876_

————————————————
_"Thank you, its now getting "failure(SwiftyInsta.GenericError.custom("Invalid response. 400"))". This is my viewController, could you please show me my errors in this one or if your have a working function for posting an image that you could just cut and paste to me that would help me tremendously, I have no idea where this is going wrong"_

```swift
    import UIKit
    import SwiftyInsta
    class DeploymentViewController: UIViewController {

        let handler = APIHandler()
        @IBOutlet weak var sampleImage: UIImageView!

  
            override func viewDidLoad() {
                    super.viewDidLoad()

            }
    
       @IBAction func test(_ sender: Any) {


            guard let key = UserDefaults.standard.string(forKey: "current.account") else { return 
    print("`key` not found.") }

            guard let cache = Authentication.Response.persisted(with: key) else { return 
    print("`Authentication.Response` not found.") }
                handler.authenticate(with: .cache(cache)) { _ in    }
        
            let img1: Upload.Picture = Upload.Picture(image: self.sampleImage.image!, caption: 
    "sampleText", size: CGSize(width: 1080,height: 1080) )
        
    
                self.handler.media.upload(photo: img1, completionHandler: { (result) in
            
            
              
                print(result)
            })    }
    

    }
```

**Comment by sbertix** [comment]
Have you tried waiting for the authentication handler to be called? @trentona
Refactoring the code Inside of `test` to…
```swift
[…]
handler.authenticate(with: .cache(cache)) { [unowned self] _ in
    // I'm assuming it logged in as planned.
    let img = Upload.Picture(image: self.sampleImage.image!, caption: 
    "sampleText", size: CGSize(width: 1080,height: 1080))
    self.handler.media.upload(photo: img1, completionHandler: { print($0) })
}
```

**Comment by trentona** [comment]
Do you have a full View Controller as an example? Im now getting failure(SwiftyInsta.GenericError.custom("Invalid response. 403"))
Iv tried a bunch of different ways and im either getting 400 or 403. Fetching the users media works fine but if I try to apply the handler in the same way I get 403. Thanks btw

**Comment by trentona** [comment]
Now it says this, 
      failure(Swift.DecodingError.typeMismatch(Foundation.Data, 
     Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "media", intValue: nil)], 
     debugDescription: "Expected to decode Data but found a dictionary instead.", underlyingError: 
     nil)))

But im giving it an image forsure,
         let sampleImage: UIImage = myImage.image!

I ended up rearranging how I had the handler to earlier and it worked. But now its mad and says it wants data but it found dictionary, my url Im assuming, that im loading the image with via kingfisher. But since im referencing sampleImage, and I made the new ref to make sure it existed, and its not crashing and the image shows up in my view controller. It wont accept pngdata or anything else i try to give it either. Why would it say im not giving it an image? Does it want the image in a specific format?

**Comment by sbertix** [comment]
I'm sorry I didn't have time to check this out @trentona 
Did you have any updates on this? @TheM4hd1 :blush:

**Comment by trentona** [comment]
@sbertix With the print out I still get that previous failure line, but the image does upload. So I it works but it says it doesn't haha. Maybe im printing it out too early or something but either way, thanks for your guys help

**Comment by sbertix** [comment]
Well… if it's a timing problem, I suggest making sure you call the `upload` method, *after* the `completionHandler` of the `authenticate` one has been called (like I did in the snippet I posted a few comments ago).
Other than that, I'm glad everything's working fine now 😊 @trentona

**Comment by mehroozkhan** [comment]
@sbertix Image does upload but getting decoding error from library : 
typeMismatch(Foundation.Data, Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "media", intValue: nil)], debugDescription: "Expected to decode Data but found a dictionary instead.", underlyingError: nil))

**Comment by sbertix** [comment]
I think I've pinpointed this new error, it has to do with `Upload.Response.Image` trying to decode `Media` the `ParsedResponse` way, instead of the regular `Decodable` way.
I'm pushing a fix together with the `2.2` PR #96.
Thx 😊

**Comment by mehroozkhan** [comment]
@sbertix thank you so much for the quick reply, Please notify me when its done.

**Comment by mehroozkhan** [comment]
@sbertix getting same error while uploading video, 
Any Idea when this will be fixed ?

**Comment by sbertix** [comment]
@mehroozkhan it's gonna be fixed with `2.2`.
Follow #96 for more info 😊
